oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+to be spec-compliant.
+
+`type <https://github.com/cloudevents/spec/blob/master/spec.md#type>`_
+----------------------------------------------------------------------
+Example: ``"org.openedx.catalog.course.created.v1"``
+
+This is the name of our event. CloudEvents only requires that this be a string
+describing the type of event, but recommends that it be prefixed with a
+reverse-DNS name for namespacing purposes. We will use a period-separated
+hierarchical name comprising with the format ``{Reverse DNS}-{Architecture
+Domain}-{Subject}-{Action}-{Major Version}``.
+
+Reverse DNS
+~~~~~~~~~~~
+Example: ``org.openedx``
+
+In this example, ``org.openedx`` is the generic prefix that would be used for
+all Open edX installs. Events that are strictly edX-specific, like those that
+might interact with internal IT or finance reporting systems, should use
+``org.edx`` as the prefix instead. If in doubt, default to ``org.openedx``.
+
+Domain (from Domain Driven Design)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Example: ``catalog``
+
+The next part of the ``type`` hierarchy is the Domain. Examples of this are:
+
+* catalog
+* content_authoring
+* credentials
+* ecommerce
+* learning
+
+It is the expectation that there are relatively few domains, and that they will
+roughly match deployed services. Domain names should be lower cased and use
+underscores if they are more than one word.
+
+Subject
+~~~~~~~
+Example: ``course``
+
+The name of an entity that the event applies to. Examples might be ``course``,
+``student``, ``enrollment``, ``order``, etc. Subjects may be namespaced, so
+``special_exam.proctored.allowance`` could be a subject.
+
+A subject should always mean the same thing within a Domain, but can mean
+different things across domains. For instance, what the LMS (``learning``
+domain) calls a ``course`` might map to what the ``catalog`` domain would call a
+``course_run``. We should try to be consistent where possible, but each domain
+ultimately gets to decide what its terms mean, and we should be careful when
+translating a concept from one domain to another. For instance, the
+``content_authoring`` and ``learning`` domains might both have a concept of a
+"due date" for an assignment. But while the ``content_authoring`` due date is
+determined only by the content author, the ``learning`` due date might take into
+account a student's cohort, individual due date extensions, accessibility
+allowances, and any number of other things. Both domains may call it ``due``,
+but the due date information from ``content_authoring`` is just an input to the
+more complex due date information in ``learning``.
+
+Action
+~~~~~~
+Example: ``created``
+
+This is the action that occurred for the event. Some of most common ones will be
+``created``, ``updated``, and ``deleted``, but many applications will want more
+specific actions like ``declined``, ``started``, ``verified``, etc. Applications
```

Comment:
🤔  -- Say I have some new python microservice and I want to start consuming messages from the LMS, I could probably have some code in a consumers file like:

```
if message.get('action') == 'created':
    run_my_ingest_logic()
elif message.get('action') == 'update':
    run_my_update_logic()
else message.get('action') == 'deleted':
    run_my_delete_logic()
```

How will I know what 'action' verbs messages from a particular service can include? Are they easily discoverable? Do I need to look at the source code of the producing service?

If we have a set vocabulary (even if only created, updated, deleted), it does simplify what the consumer needs to know about implementation details of the producing service. Granted, 90% of things are probably updates, but perhaps knowing `updated` + the subject and domain might be enough info to write your consumer code to do the thing it wants

Thoughts?

Reply:
Unpacking a few things...

**Dispatching by Event Type**

I think we're likely to have some kind of simple registration scheme where you can specify what kind of message you want, like:

```python
@handle_event("org.openedx.catalog.course.created.v1")
def do_something(event):
    # ...
```

**Knowing what a Service Offers**

This is something that we're going to have to fill with docs, but there are potentially going to be a slew of possible message types at different minor versions throughout the system. So I do think we also need to be able to generate a list of them, e.g. via some management command. I'm still fuzzy on the details, will try to add that to this OEP in the next week. I'd like to create a fairly small library, probably based on `attrs` that lets you cleanly separate minor versions and run basic sorts of introspection for reporting purposes. I'm not sure how to tie it to documentation at the moment (AsyncAPI?).

Reply:
Question that came to mind: what happens when we decide to change the action of a message that is already being consumed by others. We register a new message version right?

If so we'll definitely want to stress the importance to people to get their action verb right the first time.